core/translate.c

```diff
@@ -189,11 +189,35 @@ translate_walk_track(dcontext_t *tdcontext, instr_t *inst, translate_walk_t *wal
         walk->unsupported_mangle = false;
         walk->xsp_adjust = 0;
         for (r = 0; r < REG_SPILL_NUM; r++) {
+#ifndef ARM
             /* we should have seen a restore for every spill, unless at
              * fragment-ending jump to ibl, which shouldn't come here
              */
             ASSERT(!walk->reg_spilled[r]);
             walk->reg_spilled[r] = false; /* be paranoid */
+#else
+            /* On ARM we do spill registers across app instrs and mangle
+             * regions, though right now only the following routines do this:
+             * - mangle_stolen_reg()
+             * - mangle_gpr_list_read()
+             * - mangle_reads_thread_register()
+             * Each of these cases is a tls restore, and we assert as much.
+             */
+            DOCHECK(1, {
```

Comment:
This was done to address some pre-existing problems I saw, documented [here](https://groups.google.com/forum/#!forum/dynamorio-users).

Some questions:

* Should I make this a separate CL?
* Should we lift all this out of the `DOCHECK`?
* Is an instrlist walk too expensive, even for -debug?
* Are these three mangle functions really the only ones that reserve over an app instruction? I didn't audit all the code thoroughly, just checked for comments noting "app instruction executes here..."

PTAL